ROCm: enable cpp_extensions.load/load_inline

t-vi opened the pull request:
This enables cpp_extensions.load/load_inline. This works by hipify-ing cuda sources.
Also enable tests.
CuDNN/MIOpen extensions aren't yet supported, I propose to not do this in this PR.

t-vi committed:
ROCm: enable cpp_extensions.load/load_inline

t-vi commented:
From what I can see, the errors seem to be mostly network connectivity rather than what happens in the relevant tests.

t-vi committed:
rebuild

t-vi committed:
clean up generated files

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1056,6 +1058,20 @@ def _jit_compile(name,
+    if with_cuda is None:
+        with_cuda = any(map(_is_cuda_file, sources))
+    with_cudnn = any(['cudnn' in f for f in extra_ldflags or []])
+    if IS_HIP_EXTENSION and with_cuda or with_cudnn:
```
Did you mean `IS_HIP_EXTENSION and (with_cuda or with_cudnn)`? Because that's not what you've written here ;)

t-vi replied:
Right. :) Thanks!

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1094,6 +1110,14 @@ def _jit_compile(name,
+            try:
+                os.rmdir(os.path.dirname(fn))
+            except IOError:
+                pass
```
Is there a reason you're rm'ing the translated files?

t-vi replied:
So this is a bit of a grey area.
For extensions built from provided source files, we have, in theory, two options where to put the translated files:

- next to the source, which is what we do,
- in the build directors (/tmp/torch_extensions). This runs into trouble when you have includes etc which now are somewhere else.
So I picked the first.

But now we have modified the source directory contents (which is a bit of an annoyance while developing for PyTorch itself, too). We could keep the generated files or not.
The CI seems to enforce that we end with a clean git checkout. So I could either remove them here or remove them in the test.
_Personally_, I'd lean towards keeping the files around. But removing them in the test would make me feel dirty, too.

If you have advice, I'd gladly follow it.

t-vi replied:
I'll tentatively clean up in the test.

ezyang commented on torch/utils/hipify/hipify_python.py:
```diff
@@ -130,16 +130,21 @@ def preprocess(
                 get_hip_file_path(filepath), result)
 
     print(bcolors.OKGREEN + "Successfully preprocessed all matching files." + bcolors.ENDC, file=sys.stderr)
+    return translated_files
```
This looks wrong

t-vi replied:
If we decide to not remove the temp files, I can drop it again.
But then returning the files you created seems a relatively benign return value.

ezyang replied:
I thought it looked wrong because you were unconditionally short-circuiting (with the return here), but there was dead code after this function.

t-vi replied:
Right. Who needs debug output. :)
This is gone now along with the return value.

ezyang commented on torch/utils/hipify/hipify_python.py:
```diff
@@ -802,9 +808,10 @@ def hipify(
     all_files = list(matched_files_iter(output_directory, includes=includes,
                                         ignores=ignores, extensions=extensions,
                                         out_of_place_only=out_of_place_only))
+    all_files += [f for f in extra_files if f not in all_files]
```
quadratic police!

t-vi replied:
He. Yeah. I figured the enumeration would be short (because usually the extra_files is empty), but I'll go for a set.

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1568,7 +1599,8 @@ def sanitize_flags(flags):
     config.append('cxx = {}'.format(compiler))
     if with_cuda:
-        config.append('nvcc = {}'.format(_join_cuda_home('bin', 'nvcc')))
+        nvcc = (_join_rocm_home('bin', 'hipcc') if IS_HIP_EXTENSION else _join_cuda_home('bin', 'nvcc'))
```
nit: just don't use a ternary here lol

ezyang reviewed (changes requested):
some minor bugs

t-vi committed:
feedback comments from Ed. Thanks! keep intermediates (but clean up in test)

t-vi commented:
Thank you for the prompt review, @ezyang !
I changed to keep the intermediates, that also gets rid of some of the other uglyness.
Fixed the other bits.

t-vi commented:
So if you're OK with keeping the generated hip (after a bit more thought, I think it is preferable), this should have the things fixed.

ezyang commented on test/test_cpp_extensions_jit.py:
```diff
@@ -118,6 +120,10 @@ def test_jit_cuda_extension(self):
+        if TEST_ROCM:
+            os.unlink('cpp_extensions/hip/hip_extension.cpp')
+            os.unlink('cpp_extensions/hip/hip_extension.hip')
+            os.rmdir('cpp_extensions/hip/')
```
Hmm, this seems very error prone (especially since people may add new tests without thinking about rocm, and they wold not be aware they have to delete the temporary files. So unfortunately I think it is probably better for the code to handle cleaning up temporaries itself.

t-vi replied:
They will be reminded by the CI, though.

ezyang replied:
Yes, but only in ROCm CI; which they're at much higher risk fo ignoring (sorry, wish it wasn't true, but that's the reality)

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1056,6 +1058,18 @@ def _jit_compile(name,
+            extra_files=[os.path.abspath(s) for s in sources],
+            show_detailed=verbose,
+            is_pytorch_extension=True,
+        )
```
Maybe a better way to structure the code here is to do something similar to how TemporaryFile does it: you have an object whose destructor will clean up the temporary object. Then you can just rely the destructor to clean things up. Or better yet, have some sort of context manager (so that you can guarantee the cleanup happens in a prompt manner)

t-vi replied:
So here is why I did come to the conclusion that leaving the generated files is better
- If something goes wrong, we might want to check them.
- If nothing goes wrong, keeping the intermediates avoids recompiles in the next run.
- To me, it seems that the context manager approach adds a lot of handing around stuff to the somewhat convoluted control flow in cpp_extension.
- Translation is only done when people use ROCm-enabled versions of PyTorch. I suspect that ROCm users will be more fond of .hip translations than, say, a non-ROCm public.

But so we can do the following:
- Wrap everything from the hipify to the ninja call in a new named context manager that will then be passed to the hipify call.
- The context manager takes a new keep_intermediates argument which defaults to False except when we are working in `load_inline` (when things happen in a temp dir, so we can leave the intermediates). This is overridable  and we don't have a new optional keep_intermediates argument.

Would that work for you?

ezyang replied:
So, I agree that wanting to check them and recompiling them are legitimate arguments. 

So here's an alternative: why don't you name the files something distinctive, and then you can just blacklist them with gitignore. Then that woudl appease the CI?

t-vi replied:
Or I could just switch the default to `keep_intermediates=True` and call the test with `False`...

ezyang replied:
that would work too

t-vi replied:
So I think we have converged. :slightly_smiling_face: 
The update now does this.

t-vi committed:
keep track of intermediates in building extensions

t-vi committed:
Merge branch 'master' into hip_cpp_extension_load

t-vi committed:
default to keeping intermediates

ezyang commented on torch/utils/hipify/hipify_python.py:
```diff
@@ -72,6 +72,39 @@ class bcolors:
+    def __init__(self, keep_intermediates=False):
+        self.keep_intermediates = keep_intermediates
+        self.files_to_clean = set()
+        self.dirs_to_clean = []
```
You never write to this member variable...

t-vi committed:
record directories to clean up

ezyang commented on torch/utils/hipify/hipify_python.py:
```diff
@@ -72,6 +72,39 @@ class bcolors:
     BOLD = '\033[1m'
     UNDERLINE = '\033[4m'
 
+class GeneratedFileCleaner:
```
The naming here is a little weird, because from the context of this script itself `hipify_python.py` these aren't intermediates at all; they're the output of the hipification process. So to me this evokes more of a, "hey, please tell me what output files you're actually going to generate here!" similar to the FileManager abstraction we have in our other codegen scripts. I don't really want to bikeshed here, but a comment at this class talking about the intent (specifically for the jit extensions case) would help future refactorers a bit, I think.

t-vi replied:
Right.

t-vi committed:
add commentary

t-vi committed:
add commentary

t-vi commented:
It the commentary OK now?

ezyang commented:
Lint.

```

  {
    path: 'torch/utils/hipify/hipify_python.py',
    start_line: 823,
    end_line: 823,
    start_column: 17,
    end_column: 17,
    annotation_level: 'failure',
    message: '[B006] Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.'
  }
```

t-vi committed:
it should be don't mutate default arguments rather than don't have mutuable default arguments but who cares about nuances

t-vi commented:
Oh. Sorry about that. :man_facepalming:
